src/server_set.rs

```diff
@@ -17,12 +17,15 @@ impl ServerSet {
         let _ = self.servers.insert(addr.clone());
     }
 
-    pub fn remove_server(&mut self, addr: &PeerInfo) {
-        self.iterators.retain(|sender| {
-            sender.unbounded_send((addr.clone(), false)).is_ok()
+    pub fn remove_server(&mut self, addr: SocketAddr) {
+        let server_info = PeerInfo::with_rand_key(addr);
```

Comment:
What's going on here? Why are we generating a random key and expecting that `PeerInfo` to be present in our `servers` set?

Reply:
Well, first of I modified `ServerSet` to hold `PeerInfo` instead of `SocketAddr`. `PeerInfo` holds both `SocketAddr` and `PublicKey` - that's the required information to securely connect to a peer. Hope this part is clear.

Then, I noticed that in some cases Crust wants to remove STUN servers by IP address - no public key given. So some changes had to be made to `ServerSet`.

`Servers` and `ServerSet` interact with each other via this `UnboundedSender<PeerInfo>`. When we either remove or add servers we send a message to this channel. And in case of removal operation, there's no public key, so I simply generate it randomly.

I guess alternatively, I could make two channels: one for "add server", another one for "remove server".

Reply:
Yeah, having to generate dummy data is a big red flag that the types are wrong somewhere. I'd replace the `(PeerInfo, bool)` in the channel with `enum { Remove(SocketAddr), Add(PeerInfo) }`. And you'll need to decide how to remove the value from the `HashSet`, or replace it with a different data structure, because when you've got there won't remove anything since the public key will be different to what's in the set.

Alternatively, rather than making this change, change the part of crust that wants to remove a server based solely on a `SocketAddr`. This is why I wanted to merge the key into `PaAddr` - if everything is encrypted then a socket address should be completely useless without its key and so it's suspicious that an address is being passed around the code without a key.

Reply:
`PeerInfo` is already hashed only by address https://github.com/ustulation/p2p/pull/56/commits/56bb83f5129733702750c550a190debdaebec678 so no more changes on that are necessary.

Well, in Crust in some places we don't have public key, e.g. https://github.com/maidsafe/crust/blob/dev/src/service.rs#L115 - there's a bootstrap server blacklist which is only IP addresses.

Reply:
> `PeerInfo` is already hashed only by address 56bb83f so no more changes on that are necessary.

I wouldn't expect this to be the case. Is this just a hack to make another hack work?

Reply:
Like, maybe `ServerSet` should hold a `HashMap<SocketAddr, PublicKey>`. Thoughts?